### PR TITLE
Feature/get protected data allow types array

### PR DESCRIPTION
**Pull request by PierreJeanjacquot**
# Features

- allow legacy types `"boolean"` and `"number"` in `getProtectedData()` `requiredSchema`

```ts
await dataprotectorCore.getProtectedData(
  requiredSchema: {
    isUsCitizen: "boolean",
    yearOfBirth: "number"
  }
);
```

- allow searching any type in list in `getProtectedData()` `requiredSchema`

```ts
await dataprotectorCore.getProtectedData(
  requiredSchema: {
    picture: ["image/png", "image/jpeg"]
  }
);
```

### CODE REVIEW

File: packages/sdk/src/lib/dataProtectorCore/getProtectedData.ts [hunk 1]
```diff
@@ -56,49 +56,63 @@ export const getProtectedData = async ({
...
+        or: anyOfSchema.map((schemaFragment) => ({
+          schema_contains: [schemaFragment],
+        })),
+      });
+    });
+
+    const filteredProtectedDataQuery = gql`
+      query ($start: Int!, $range: Int!, $where: ProtectedData_filter) {
+        protectedDatas(
+          where: $where
```
Comment (cedric25) [inline packages/sdk/src/lib/dataProtectorCore/getProtectedData.ts:85]: +1 for not having a complicated template string here 👍

File: packages/sdk/src/lib/dataProtectorCore/getProtectedData.ts [hunk 2]
```diff
@@ -109,15 +123,37 @@ export const getProtectedData = async ({
...
-    }
-  });
+function flattenSchema(
+  schema: SearchableDataSchema,
+  parentKey?: string
+): { requiredSchemas: string[]; anyOfSchemas: string[][] } {
+  return Object.entries(schema).reduce(
+    (acc, [key, value]) => {
+      const newKey = parentKey ? `${parentKey}.${key}` : key;
+      // any of types
```
Comment (cedric25) [inline packages/sdk/src/lib/dataProtectorCore/getProtectedData.ts:133]: Same comment as in other pieces of code: "any of" is not easy to understand for me.

```suggestion
      // Array of possible types
```

File: packages/sdk/src/lib/dataProtectorCore/getProtectedData.ts [hunk 3]
```diff
@@ -109,15 +123,37 @@ export const getProtectedData = async ({
...
+      if (Array.isArray(value)) {
+        if (value.length > 1) {
+          // one of many
+          acc.anyOfSchemas.push(value.map((entry) => `${newKey}:${entry}`));
+        } else {
+          // one of one
```
Comment (cedric25) [inline packages/sdk/src/lib/dataProtectorCore/getProtectedData.ts:139]:
```suggestion
          // Array of only one type. Similar to single type.
```

Is that correct to say that?

**Review by cedric25 — COMMENTED** [review]
No real blocker found on this PR, good job!
Some comments about wording.

File: packages/sdk/tests/e2e/dataProtectorCore/getProtectedData.test.ts [hunk 4]
```diff
@@ -72,52 +83,165 @@ describe('dataProtectorCore.getProtectedData()', () => {
...
+
+        // --- THEN
+        expect(result.length).toEqual(2);
+        expect(
+          result[0].name === 'bool' || result[0].name === 'f64'
+        ).toBeTruthy();
+        expect(
+          result[1].name === 'bool' || result[1].name === 'f64'
+        ).toBeTruthy();
+        expect(result[0].name !== result[1].name).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/tests/e2e/dataProtectorCore/getProtectedData.test.ts:241]: right! I initially did this and had a failed test, I'll check why, thank you for the heads up!

File: packages/sdk/tests/e2e/dataProtectorCore/getProtectedData.test.ts [hunk 5]
```diff
@@ -36,7 +33,21 @@ describe('dataProtectorCore.getProtectedData()', () => {
       });
       expect(res).toBeDefined();
     },
-    MAX_EXPECTED_WEB2_SERVICES_TIME
+    timeouts.getProtectedData
+  );
+
+  it(
+    'accept any of array in requiredSchema',
```
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/tests/e2e/dataProtectorCore/getProtectedData.test.ts:40]:
```suggestion
    'accept array of possible type in requiredSchema',
```

File: packages/sdk/tests/e2e/dataProtectorCore/getProtectedData.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (PierreJeanjacquot) [reply packages/sdk/tests/e2e/dataProtectorCore/getProtectedData.test.ts]: seems I expected asc sorting, fixed in 4f255eefc2de1cebe7e50bea97c552f931cf99c6

**Review by cedric25 — APPROVED** [review]

**Review by ErwanDecoster — APPROVED** [review]

**Pull request merged**